integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/facets/builder/DatasetVersionDatasetFacetUtils.java

```diff
@@ -0,0 +1,60 @@
+package io.openlineage.spark3.agent.facets.builder;
+
+import static io.openlineage.spark.agent.util.ScalaConversionUtils.asJavaOptional;
+
+import io.openlineage.spark3.agent.lifecycle.plan.catalog.CatalogUtils3;
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.delta.files.TahoeLogFileIndex;
+import org.apache.spark.sql.execution.datasources.HadoopFsRelation;
+import org.apache.spark.sql.execution.datasources.LogicalRelation;
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation;
+
+@Slf4j
+class DatasetVersionDatasetFacetUtils {
+
+  /**
+   * Check if we have all the necessary properties in DataSourceV2Relation to get dataset version.
+   */
+  public static Optional<String> extractVersionFromDataSourceV2Relation(
+      DataSourceV2Relation table) {
+    if (table.identifier().isEmpty()) {
+      log.warn("Couldn't find identifier for dataset in plan " + table);
+      return Optional.empty();
+    }
+    Identifier identifier = table.identifier().get();
+
+    if (table.catalog().isEmpty() || !(table.catalog().get() instanceof TableCatalog)) {
+      log.warn("Couldn't find catalog for dataset in plan " + table);
+      return Optional.empty();
+    }
+    TableCatalog tableCatalog = (TableCatalog) table.catalog().get();
+
+    Map<String, String> tableProperties = table.table().properties();
+    return CatalogUtils3.getDatasetVersion(tableCatalog, identifier, tableProperties);
+  }
+
+  /**
+   * Delta uses LogicalRelation's HadoopFsRelation as a logical plan's leaf. It implements FileIndex
+   * using TahoeLogFileIndex that contains DeltaLog, which can be used to get dataset's snapshot.
+   */
+  public static Optional<String> extractVersionFromLogicalRelation(
+      LogicalRelation logicalRelation) {
+    if (logicalRelation.relation() instanceof HadoopFsRelation) {
+      HadoopFsRelation fsRelation = (HadoopFsRelation) logicalRelation.relation();
+      asJavaOptional(logicalRelation.catalogTable());
+      if (logicalRelation.catalogTable().isDefined()
+          && logicalRelation.catalogTable().get().provider().isDefined()
+          && logicalRelation.catalogTable().get().provider().get().equalsIgnoreCase("delta")) {
```

Comment:
Is it possible to move this logic into handler classes? So far we managed to avoid any provider specific logic outside provider handlers.

Reply:
Would be done this way if this was v2 catalog 😉